Finish to/from methods for Temporal.PlainDate[Time]

rkirsling opened the pull request:
#### 67370a953ddddcf9b3331cc06bc7e0fab6f79846
<pre>
Finish to/from methods for Temporal.PlainDate[Time]
<a href="https://bugs.webkit.org/show_bug.cgi?id=244436">https://bugs.webkit.org/show_bug.cgi?id=244436</a>

Reviewed by Yusuke Suzuki.

This patch implements the following for the Temporal API:
- PlainDate: `from(object)`, toPlainDateTime
- PlainDateTime: `from(object)`, toPlainDate, toPlainTime
- PlainTime: toPlainDateTime
- Calendar: dateFromFields (where the actual implementation of `PlainDate.from(object)` lives)

It then enables test262 tests for all implemented methods of PlainDate and PlainDateTime,
and subsequently recategorizes all of our existing Temporal failures, while fixing all imminently addressable issues.

* JSTests/stress/temporal-calendar.js:
* JSTests/stress/temporal-plaindate.js:
* JSTests/stress/temporal-plaindatetime.js:
* JSTests/stress/temporal-plaintime.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::temporalDateToString): Fix six-digit year handling.
(JSC::ISO8601::monthFromCode): Added.
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::toTemporalCalendarWithISODefault): Moved from TemporalPlainTime.
(JSC::TemporalCalendar::getTemporalCalendarWithISODefault): Moved from TemporalPlainTime.
(JSC::TemporalCalendar::isoDateFromFields): Added.
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalCalendarPrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
(JSC::TemporalPlainDate::toPlainDate): Use max/min year constants to ensure that we *actually* stay within 21 bits.
(JSC::TemporalPlainDate::tryCreateIfValid):
(JSC::TemporalPlainDate::from):
* Source/JavaScriptCore/runtime/TemporalPlainDate.h: Use decltype(auto) because year is int32_t.
* Source/JavaScriptCore/runtime/TemporalPlainDateConstructor.cpp:
(JSC::TemporalPlainDateConstructor::finishCreation):
* Source/JavaScriptCore/runtime/TemporalPlainDatePrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.cpp:
(JSC::TemporalPlainDateTime::tryCreateIfValid):
(JSC::TemporalPlainDateTime::from):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.h: Use decltype(auto) because year is int32_t.
* Source/JavaScriptCore/runtime/TemporalPlainDateTimeConstructor.cpp:
(JSC::TemporalPlainDateTimeConstructor::finishCreation):
* Source/JavaScriptCore/runtime/TemporalPlainDateTimePrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
(JSC::TemporalPlainTime::toTemporalTimeRecord): Made public static.
(JSC::TemporalPlainTime::regulateTime): Made public static.
(JSC::TemporalPlainTime::from):
(JSC::toTemporalCalendarWithISODefault): Moved to TemporalPlainTime.
(JSC::getTemporalCalendarWithISODefault): Moved to TemporalPlainTime.
* Source/JavaScriptCore/runtime/TemporalPlainTime.h:
* Source/JavaScriptCore/runtime/TemporalPlainTimePrototype.cpp:

Canonical link: <a href="https://commits.webkit.org/253876@main">https://commits.webkit.org/253876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9a8b21298d4b7de105aadd338c17165d757720d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87344 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96573 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149919 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29796 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25995 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79461 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91346 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74136 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23914 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79267 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66930 "Found 1 new API test failure: /WebKit2Gtk/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79196 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27512 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13102 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72877 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27464 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14117 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25953 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2727 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29151 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36979 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75697 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29082 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33394 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16768 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -114,4 +143,90 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        return { };
+    }
+
+    JSValue monthProperty = temporalDateLike->get(globalObject, vm.propertyNames->month);
```
Need exception check here.

rkirsling replied:
Oops, thanks!

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -114,4 +143,90 @@ JSObject* TemporalCalendar::from(JSGlobalObject* globalObject, JSValue calendarL
+        RETURN_IF_EXCEPTION(scope, { });
+    }
+
+    JSValue monthCodeProperty = temporalDateLike->get(globalObject, vm.propertyNames->monthCode);
```
Ditto.

Constellation reviewed (approved):
r=me

rkirsling committed:
Finish to/from methods for Temporal.PlainDate[Time]
https://bugs.webkit.org/show_bug.cgi?id=244436

Reviewed by Yusuke Suzuki.

This patch implements the following for the Temporal API:
- PlainDate: `from(object)`, toPlainDateTime
- PlainDateTime: `from(object)`, toPlainDate, toPlainTime
- PlainTime: toPlainDateTime
- Calendar: dateFromFields (where the actual implementation of `PlainDate.from(object)` lives)

It then enables test262 tests for all implemented methods of PlainDate and PlainDateTime,
and subsequently recategorizes all of our existing Temporal failures, while fixing all imminently addressable issues.

* JSTests/stress/temporal-calendar.js:
* JSTests/stress/temporal-plaindate.js:
* JSTests/stress/temporal-plaindatetime.js:
* JSTests/stress/temporal-plaintime.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::temporalDateToString): Fix six-digit year handling.
(JSC::ISO8601::monthFromCode): Added.
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::toTemporalCalendarWithISODefault): Moved from TemporalPlainTime.
(JSC::TemporalCalendar::getTemporalCalendarWithISODefault): Moved from TemporalPlainTime.
(JSC::TemporalCalendar::isoDateFromFields): Added.
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalCalendarPrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp:
(JSC::TemporalPlainDate::toPlainDate): Use max/min year constants to ensure that we *actually* stay within 21 bits.
(JSC::TemporalPlainDate::tryCreateIfValid):
(JSC::TemporalPlainDate::from):
* Source/JavaScriptCore/runtime/TemporalPlainDate.h: Use decltype(auto) because year is int32_t.
* Source/JavaScriptCore/runtime/TemporalPlainDateConstructor.cpp:
(JSC::TemporalPlainDateConstructor::finishCreation):
* Source/JavaScriptCore/runtime/TemporalPlainDatePrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.cpp:
(JSC::TemporalPlainDateTime::tryCreateIfValid):
(JSC::TemporalPlainDateTime::from):
* Source/JavaScriptCore/runtime/TemporalPlainDateTime.h: Use decltype(auto) because year is int32_t.
* Source/JavaScriptCore/runtime/TemporalPlainDateTimeConstructor.cpp:
(JSC::TemporalPlainDateTimeConstructor::finishCreation):
* Source/JavaScriptCore/runtime/TemporalPlainDateTimePrototype.cpp:
* Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
(JSC::TemporalPlainTime::toTemporalTimeRecord): Made public static.
(JSC::TemporalPlainTime::regulateTime): Made public static.
(JSC::TemporalPlainTime::from):
(JSC::toTemporalCalendarWithISODefault): Moved to TemporalPlainTime.
(JSC::getTemporalCalendarWithISODefault): Moved to TemporalPlainTime.
* Source/JavaScriptCore/runtime/TemporalPlainTime.h:
* Source/JavaScriptCore/runtime/TemporalPlainTimePrototype.cpp:

Canonical link: https://commits.webkit.org/253876@main
